File: src/main/java/sunshine/infrastructure/configuration/RetryInterceptor.java
package sunshine.infrastructure.configuration;

import java.io.IOException;
import org.springframework.http.HttpRequest;
import org.springframework.http.client.ClientHttpRequestExecution;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.http.client.ClientHttpResponse;

public class RetryInterceptor implements ClientHttpRequestInterceptor {

  private final int maxAttempts;
  private final long backoff;

  public RetryInterceptor(int maxAttempts, long backoff) {
    this.maxAttempts = maxAttempts;
    this.backoff = backoff;
  }

  @Override
  public ClientHttpResponse intercept(
      HttpRequest request,
      byte[] body,
      ClientHttpRequestExecution execution
  ) throws IOException {
    return executeWithRetry(request, body, execution, 0);
  }

  private ClientHttpResponse executeWithRetry(
      HttpRequest request,
      byte[] body,
      ClientHttpRequestExecution execution,
      int attempt
  ) throws IOException {
    try {
      return execution.execute(request, body);
    } catch (IOException e) {
      if (attempt >= maxAttempts - 1) {
        throw e;
      }
      try {
        Thread.sleep(backoff);
      } catch (InterruptedException ie) {
        Thread.currentThread().interrupt();
        throw new IOException("Interrupted during retry backoff", ie);
      }
      return executeWithRetry(request, body, execution, attempt + 1);
    }
  }
}

File: src/main/kotlin/sunshine/infrastructure/configuration/RetryInterceptor.kt
package sunshine.infrastructure.configuration

import org.springframework.http.HttpRequest
import org.springframework.http.client.ClientHttpRequestExecution
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.http.client.ClientHttpResponse
import java.io.IOException

class RetryInterceptor(
    private val maxAttempts: Int,
    private val backoff: Long
) : ClientHttpRequestInterceptor {

    override fun intercept(
        request: HttpRequest,
        body: ByteArray,
        execution: ClientHttpRequestExecution
    ): ClientHttpResponse {
        return executeWithRetry(request, body, execution, 0)
    }

    private fun executeWithRetry(
        request: HttpRequest,
        body: ByteArray,
        execution: ClientHttpRequestExecution,
        attempt: Int
    ): ClientHttpResponse {
        try {
            return execution.execute(request, body)
        } catch (e: IOException) {
            return handleException(request, body, execution, attempt, e)
        }
    }

    private fun handleException(
        request: HttpRequest,
        body: ByteArray,
        execution: ClientHttpRequestExecution,
        attempt: Int,
        e: IOException
    ): ClientHttpResponse {
        if (attempt >= maxAttempts - 1) {
            throw e
        }
        Thread.sleep(backoff)
        return executeWithRetry(request, body, execution, attempt + 1)
    }
}
